### ISSUE TITLE
net/tstun: failed to set MTU of TUN device: invalid argument

### ISSUE THREAD

**Issue by bradfitz** [tailscale/tailscale#11899]
\[ copying from https://www.reddit.com/r/Tailscale/comments/1cegdm2/error_with_the_static_arm_build_1582/ ]

> I use the Static ARM build of tailscale on a Seagate GoFlex Home NAS Device, which is running Debian Linux.
>
> Up until version 1.58.2 everything has been working flawlessly.. however on any version above that I get an error when trying too bring up tailscale.
```
Linux kernel version: 5.13.6-kirkwood-tld-1
'modprobe tun' successful
/dev/net/tun: Dcrw-rw-rw-
wgengine.NewUserspaceEngine(tun "tailscale0") error: tstun.New("tailscale0"): failed to set MTU of TUN device: invalid argument
flushing log.
logger closing down
getLocalBackend error: createEngine: tstun.New("tailscale0"): failed to set MTU of TUN device: invalid argument
```
> Reverting (copying the static binaries from 1.58.2 back over top of the newer ones).. fixes the issue every time.
>
> Can anyone advise how I can get more detail about whats going wrong here?

**Comment by sambartle** [comment]
Kernel Config attached. 
[config-5.13.6-kirkwood-tld-1.txt](https://github.com/tailscale/tailscale/files/15139547/config-5.13.6-kirkwood-tld-1.txt)
(I Realised I had a tarball of the rootfs on my machine)

**Comment by bradfitz** [comment]
Any chance you could git bisect and find the commit where it fails to start?

**Comment by sambartle** [comment]
Sure, I'll have a look tomorrow morning (I'm in the UK) and figure it out.

**Comment by bradfitz** [comment]
I looked at the diff from 1.58.2 to 1.60.0. Not a ton in there. Go 1.22 instead of Go 1.21, and bumps from golang.org/x/sys 0.15 to 0.16. One of those seems the most likely candidate.

Building 1.58.2 but with Go 1.22 and/or golang.org/x/sys 0.16 would be interesting.

Another thing that would be interesting would be `strace -f tailscaled` output to see the exact system calls that are failing
.

**Comment by sambartle** [comment]
Checking out 1.58.2 and modifying it to build with Go 1.22.0 causes the same error once built (It works when built as is with 1.21 as a sanity (first time I've used Go) on the same system)..

The problem is I'm unfortunately definitely no Go expert to diagnose further.  I'll try and see if I can build a newer version with the older Go and see what happens and do some learning.

**Comment by bradfitz** [comment]
Looks like it's:

```
4081  ioctl(15, SIOCSIFMTU, {ifr_name="tailscale0", ifr_mtu=0} <unfinished ...>
...
4081  <... ioctl resumed>)              = -1 EINVAL (Invalid argument)
```

And what's the strace look like when it works?

**Comment by sambartle** [comment]
Apologies I intended to attach both.. 
[strace.txt](https://github.com/tailscale/tailscale/files/15142721/strace.txt)

4595  ioctl(15, SIOCSIFMTU, {ifr_name="tailscale0", ifr_mtu=1280} <unfinished ...>
4591  nanosleep({tv_sec=0, tv_nsec=20000},  <unfinished ...>
4595  <... ioctl resumed>)              = 0
4595  close(15 <unfinished ...>

So that definitly looks like a possibility.

**Comment by bradfitz** [comment]
@sambartle, can you try building this tiny Go program on your device, building it with both Go 1.21 and Go 1.22?

[mtu.tar.gz](https://github.com/tailscale/tailscale/files/15152945/mtu.tar.gz)

To build, from that directory:

```
$ GOOS=linux GOARCH=arm /path/to/go1.22/bin/go build -o mtu122
$ GOOS=linux GOARCH=arm /path/to/go1.21/bin/go build -o mtu121
```

Then running it on the device, pick some existing interface and run like:

```
nas$ sudo ./mtu122 --name=ens18 --mtu=1500
nas$ sudo ./mtu122 --name=ens18 --mtu=1499
nas$ sudo ./mtu122 --name=ens18 --mtu=1500
```

(and check `ifconfig` or `ip` or whatever to see that it worked?)

I'm curious if from that minimal repro we can see a difference on your device.

**Comment by bradfitz** [comment]
For the record, that code is from wireguard-go, which does:

```go
func (tun *NativeTun) setMTU(n int) error {
	name, err := tun.Name()
	if err != nil {
		return err
	}

	// open datagram socket
	fd, err := unix.Socket(
		unix.AF_INET,
		unix.SOCK_DGRAM|unix.SOCK_CLOEXEC,
		0,
	)
	if err != nil {
		return err
	}

	defer unix.Close(fd)

	// do ioctl call
	var ifr [ifReqSize]byte
	copy(ifr[:], name)
	*(*uint32)(unsafe.Pointer(&ifr[unix.IFNAMSIZ])) = uint32(n)
	_, _, errno := unix.Syscall(
		unix.SYS_IOCTL,
		uintptr(fd),
		uintptr(unix.SIOCSIFMTU),
		uintptr(unsafe.Pointer(&ifr[0])),
	)

	if errno != 0 {
		return fmt.Errorf("failed to set MTU of TUN device: %w", errno)
	}

	return nil
}
```

**Comment by sambartle** [comment]
Apologies for the delay in taking a look at this (and thanks for the ongoing help)..

When built with go 1.21 your sample code behaves as expected and sets the mtu. 
With 1.22, the code completes successfully (errno: 0 same as with 1.21), but the mtu is set to 0 and not the requested value.

**Comment by SoftwareEngineerChris** [comment]
For anyone else looking for a workaround, this issue doesn't happen when using userspace networking. However that brings problems of its own. Any update on a fix for this one? Thanks.

**Comment by bradfitz** [comment]
@sambartle, it's hard for me to debug this without a local way to repro.

If I had such a device/CPU/OS combination that was affected I'd try to bisect which Go commit caused the regression.

What's the exact model of your device? I can try to find one on Ebay, as it seems like it's been discontinued?

I'll also file an upstream Go bug and see if anybody has any ideas.

Can you attach binaries of both mtu122 and mtu121 as you compiled them, as well as run `sudo strace -o trace122.txt -f ./mtu122 --...` and attach the trace122.txt and trace121.txt files?

**Comment by bradfitz** [comment]
@sambartle, on a whim, can you also try changing the line:

```
*(*uint32)(unsafe.Pointer(&ifr[ifNamSize])) = uint32(n)
```
to instead be:
```
binary.NativeEndian.PutUint32(ifr[ifNamSize:], uint32(n))
```

Maybe that's where the bad codegen/alignment is.

**Comment by sambartle** [comment]
@bradfitzOf course I appreciate its impossible without the device in front of you.. I have the same issue as its in my office (which I'm never in) and tailscale is the way I access it.. So when I break tailscale that causes some issues reconnecting, and I have to get someone to intervene for me.. I'm planning to bring it home when I'm next in and then I can see what I can do. 

Its an old Seagate GoFlex Home NAS, however it uses a generic Marvell Kirkwood SOC which was used in all sorts of appliances:
GoFlex Home/Net, Pogoplug E02/Mobile/V4, iConnect, Dockstar, Sheevaplug, NSA320, NSA320S, NSA325, NSA310S, NSA310, Topkick, Netgear Stora, ASUS-M25, Iomega ix4-200d, QiZhiTong 501mv2, HP Thin Client T5325, LG N1T1, Linksys EA4500, Lenovo/Iomega ix2-ng, Toshiba Canvio Home NAS, Lenovo Iomega EZ, Dell KACE M300, Freecom Silverstore HNCN2, DNS-320-B1, Medion P89634, Netgear Ready NAS Duo v2, Lenovo/Iomega ix2-dl, CTera C200 V1, Synology DS413j (work in progress), Check Point L-50, Synology DS411jv10, all other Kirkwood Plugs

The kernel and most of the work getting it too boot was from here: https://forum.doozan.com/read.php?2,12096

If I bring it home I can try the updated kernel and see if that works as mine is very outdated. 

mtu binaries and strace attached.. 
[trace122.txt](https://github.com/tailscale/tailscale/files/15503929/trace122.txt)
[binaries.zip](https://github.com/tailscale/tailscale/files/15503934/binaries.zip)

**Comment by sambartle** [comment]
> @sambartle, on a whim, can you also try changing the line:
> 
> ```
> *(*uint32)(unsafe.Pointer(&ifr[ifNamSize])) = uint32(n)
> ```
> 
> to instead be:
> 
> ```
> binary.NativeEndian.PutUint32(ifr[ifNamSize:], uint32(n))
> ```
> 
> Maybe that's where the bad codegen/alignment is.

I get the same result as before (invalid argument) after making that change and compiling with go 1.22

**Comment by Fvdwnl** [comment]
I had the same problem on a kirkwood based device. The function setting the MTU value is same as in previous version and it is also used to set tun device name which succeeds. Must be something in go 23 (and 22) that it doesn't compile it correctly. Anyhow I solved it by replacing the socket setting call in that function  by a system function call to the linux ip command to set mtu valuein that routine. It is a work around but works well. tailscaled is now running just fine with correct mtu value.

**Comment by bradfitz** [comment]
I tried to repro on an old DS212j I had lying around, but it worked fine. My little test program worked fine even with recent versions of Go. The DS212j is a Marvell Kirkwood 88F6281, not the Marvell Kirkwood 88F6282 that's in the DS413j and perhaps your devices.

**Comment by bradfitz** [comment]
@Fvdwnl, @sambartle, can one (or both) of you try again using this version?

[mtu-v2.tar.gz](https://github.com/user-attachments/files/17724350/mtu-v2.tar.gz)

Try it both with and without the `-xsys` flag and see if that makes a difference. (adjust the interface name as needed)

**Comment by bradfitz** [comment]
I sent https://github.com/tailscale/wireguard-go/pull/34 on the theory that it can't be worse than the current situation and it might even fix it.

**Comment by Fvdwnl** [comment]
> I tried to repro on an old DS212j I had lying around, but it worked fine. My little test program worked fine even with recent versions of Go. The DS212j is a Marvell Kirkwood 88F6281, not the Marvell Kirkwood 88F6282 that's in the DS413j and perhaps your devices.

I have the problem with both (lacie 2bignas having 6281 and 5big2 with 6182 cpu.
With my "patch" I get this output
`dns: [rc=unknown ret=direct]
dns: using "direct" mode
dns: using *dns.directManager
cleanup: list tables: netlink receive: invalid argument
wgengine.NewUserspaceEngine(tun "tailscale0") ...
set MTU of TUN device tailscale0 MTU = 1280
command to set mtu : ip link set dev tailscale0 mtu 1280
setting link attributes: setsockopt: protocol not available
router: error querying IP rules (does kernel have IP_MULTIPLE_TABLES?): address family not supported by protocol
router: warning: running without policy routing`

notice the message setting link attributes: setsockopt: protocol not available.
Could this be pointer where the problem might be.

For info this my patch code
`func (tun *NativeTun) setMTU(n int) error {
	name, err := tun.Name()
	if err != nil {
		return err
	}
//fvdw
fmt.Printf("set MTU of TUN device %s MTU = %d\n",name, n)
const ShellToUse = "bash"
var stdout bytes.Buffer
arg0 := fmt.Sprintf("ip link set dev %s mtu %d", name, n) 
fmt.Printf("command to set mtu : %s\n", arg0)
cmd := exec.Command(ShellToUse,"-c",arg0)
cmd.Stdout = &stdout
err1 := cmd.Run()
if err != nil {
   fmt.Printf("cmd to set mtu failed err1=%d\n",err1)
}
//fvdw this function doesn't work currently on kirkwood
	// open datagram socket
/*
*	fd, err := unix.Socket(
*		unix.AF_INET,
*		unix.SOCK_DGRAM|unix.SOCK_CLOEXEC,
*		0,
*	)
*	if err != nil {
*		return err
*	}
*
*	defer unix.Close(fd)
*
*	// do ioctl call
*	var ifr [ifReqSize]byte
*	copy(ifr[:], name)
*	*(*uint32)(unsafe.Pointer(&ifr[unix.IFNAMSIZ])) = uint32(n)
*	_, _, errno := unix.Syscall(
*		unix.SYS_IOCTL,
*		uintptr(fd),
*		uintptr(unix.SIOCSIFMTU),
*		uintptr(unsafe.Pointer(&ifr[0])),
*	)
*
*	if errno != 0 {
*		return fmt.Errorf("failed to set MTU of TUN device: %w", errno)
*	}
*/
	return nil
}`

I have very limted knowledge and experience with go but I will try to compile the little program provided by bradfitz to test.

**Comment by Fvdwnl** [comment]
I compiled the program with go 23 and this is the result
```
root@fvdwsl-5big2:/share/1000/fvdw/2-test/mtu # ./mtu.arm5 --name=tailscale0 --mtu=1400
2024/11/13 15:11:33 errno: invalid argument
root@fvdwsl-5big2:/share/1000/fvdw/2-test/mtu #
""

**Comment by bradfitz** [comment]
Great! It works. We can consider this issue closed. This'll be fixed in Tailscale 1.78.0.  I'm not sure whether we'll backport it to 1.76.x yet.

**Comment by bradfitz** [comment]
Thanks for testing, @Fvdwnl!
